src/core/math_func.hpp

```diff
@@ -352,4 +352,34 @@ constexpr int RoundDivSU(int a, uint b)
 
 uint32_t IntSqrt(uint32_t num);
 
+/**
+ * Scale a number by the required percentage.
+ *
+ * Calculation is performed in the type U of the num parameter.
+ * The result is clamped to the limits of type T if needed.
+ *
+ * @param num The number to scale.
+ * @param percentage The percentage value. 100% = don't scale.
+ * @return The number scaled by the percentage value.
+ */
+template <typename T, typename U>
+constexpr T ScaleByPercentage(U num, uint16_t percentage)
+{
+	U scaled;
+	/* We might not need to do anything. */
+	if (num == 0 || percentage == 100) {
+		scaled = num;
+	} else {
+		scaled = (num * static_cast<U>(percentage)) / 100;
+	}
+
+	/* If there is no type conversion we're done. */
+	if constexpr (std::is_same_v<T, U>) {
+		return scaled;
+	}
+
+	/* Make sure the value fits resulting type T (clapmping the ends). */
+	return static_cast<T>(Clamp<U>(scaled, std::numeric_limits<T>::min(), std::numeric_limits<T>::max()));
```

Comment:
We already have ClampTo which handles edge cases, is that usable here instead?

Reply:
oh, yeah, missed that